src/hotspot/share/utilities/exceptions.cpp

```diff
@@ -262,8 +263,24 @@ void Exceptions::fthrow(JavaThread* thread, const char* file, int line, Symbol*
   va_list ap;
   va_start(ap, format);
   char msg[max_msg_size];
-  os::vsnprintf(msg, max_msg_size, format, ap);
+  int ret = os::vsnprintf(msg, max_msg_size, format, ap);
   va_end(ap);
+  // If ret == -1 then either there was a format conversion error, or the required buffer size
+  // exceeds INT_MAX and so couldn't be returned (undocumented behaviour of vsnprintf). Depending
+  // on the platform the buffer may be filled to its capacity (Linux), filled to the conversion
+  // that encountered the overflow (macOS), or is empty (Windows), so it is possible we
+  // have a truncated UTF-8 sequence. Similarly, if the buffer was too small and ret >= max_msg_size
+  // we may also have a truncated UTF-8 sequence. In such cases we need to fix the buffer so the UTF-8
+  // sequence is valid.
+  if ((ret == -1 || ret >= max_msg_size) && strlen(msg) > 0) {
+    assert(msg[max_msg_size - 1] == '\0', "should be null terminated");
```

Comment:
Would this always be true? For a formatting error, too?
Maybe just to be sure, instead of asserting set the last byte to zero.

Reply:
vsnprintf is supposed to guarantee it, and os::vsnprint does IIRC, so this is just a sanity check.

Reply:
Yep os::vnsprintf guarantees nul-termination